src/coord/src/persistence.rs

```diff
@@ -150,14 +164,43 @@ impl Source {
                 // We'll write down the data to a file with a `-tmp` prefix to
                 // indicate a write was in progress, and then atomically rename
                 // when we are done to indicate the write is complete.
-                let tmp_path = self.path.join(format!("{}-tmp", filename));
+                let tmp_path = self.path.join(format!("{}-tmp", file_name));
+                let path = self.path.join(file_name);
 
-                std::fs::write(&tmp_path, buf)?;
-                let final_path = self.path.join(filename);
-                std::fs::rename(tmp_path, final_path)?;
+                tokio::spawn(Self::flush(path, tmp_path, buf));
                 partition.last_persisted_offset = Some(prefix_end_offset);
+                persisted_records += prefix_length;
             }
         }
+
+        assert!(
+            persisted_records <= self.pending_records,
+            "persisted {} records but only had {} pending",
+            persisted_records,
+            self.pending_records
+        );
+
+        self.pending_records -= persisted_records;
+
+        // TODO(rkhaitan): Reevaluate this. On the one hand, we need to have something like this to make sure
+        // we don't get spammed into trying to write every time on a topic with a missing offset / extreme
+        // out of order-ness. On the other hand, this error is extremely opaque for the end user, and it's
+        // unclear what a user could do to resolve it.
+        if self.pending_records > self.max_pending_records {
+            anyhow!(
```

Comment:
I don't think this does anything? Did you mean to `bail!`?

Reply:
whoops yes fixed.